storage/factory/pruningStorerFactory.go

```diff
@@ -0,0 +1,537 @@
+		return nil, err
+	}
+	successfullyCreatedStorers = append(successfullyCreatedStorers, heartbeatStorageUnit)
+
+	bootstrapUnitArgs := &pruning.PruningStorerArgs{
+		Identifier:            "bootstrapUnit",
+		FullArchive:           fullArchiveMode,
+		CacheConf:             GetCacherFromConfig(psf.generalConfig.BootstrapStorage.Cache),
+		DbPath:                filepath.Join(psf.uniqueID, psf.generalConfig.BootstrapStorage.DB.FilePath),
+		PersisterFactory:      NewPersisterFactory(psf.generalConfig.BootstrapStorage.DB),
+		BloomFilterConf:       GetBloomFromConfig(psf.generalConfig.BootstrapStorage.Bloom),
+		NumOfEpochsToKeep:     numOfEpochsToKeep,
+		NumOfActivePersisters: numOfActivePersisters,
+		Notifier:              psf.epochStartNotifier,
+	}
+	bootstrapUnit, err = pruning.NewPruningStorer(bootstrapUnitArgs)
+	if err != nil {
+		return nil, err
+	}
+	successfullyCreatedStorers = append(successfullyCreatedStorers, bootstrapUnit)
+
+	store := dataRetriever.NewChainStorer()
+	store.AddStorer(dataRetriever.TransactionUnit, txUnit)
+	store.AddStorer(dataRetriever.MiniBlockUnit, miniBlockUnit)
+	store.AddStorer(dataRetriever.PeerChangesUnit, peerBlockUnit)
+	store.AddStorer(dataRetriever.BlockHeaderUnit, headerUnit)
+	store.AddStorer(dataRetriever.MetaBlockUnit, metachainHeaderUnit)
+	store.AddStorer(dataRetriever.UnsignedTransactionUnit, unsignedTxUnit)
+	store.AddStorer(dataRetriever.RewardTransactionUnit, rewardTxUnit)
+	store.AddStorer(dataRetriever.MetaHdrNonceHashDataUnit, metaHdrHashNonceUnit)
+	hdrNonceHashDataUnit := dataRetriever.ShardHdrNonceHashDataUnit + dataRetriever.UnitType(psf.shardCoordinator.SelfId())
+	store.AddStorer(hdrNonceHashDataUnit, shardHdrHashNonceUnit)
+	store.AddStorer(dataRetriever.HeartbeatUnit, heartbeatStorageUnit)
+	store.AddStorer(dataRetriever.BootstrapUnit, bootstrapUnit)
+
+	return store, err
+}
+
+// CreateForMeta will return the storage service which contains all storers needed for metachain
+func (psf *StorageServiceFactory) CreateForMeta() (dataRetriever.StorageService, error) {
+	var shardDataUnit *pruning.PruningStorer
+	var metaBlockUnit *pruning.PruningStorer
+	var headerUnit *pruning.PruningStorer
+	var txUnit *pruning.PruningStorer
+	var peerDataUnit *pruning.PruningStorer
+	var metaHdrHashNonceUnit *pruning.PruningStorer
+	var miniBlockUnit *pruning.PruningStorer
+	var unsignedTxUnit *pruning.PruningStorer
+	var miniBlockHeadersUnit *pruning.PruningStorer
+	var shardHdrHashNonceUnits []*pruning.PruningStorer
+	var bootstrapUnit *pruning.PruningStorer
+	var err error
+
+	successfullyCreatedStorers := make([]storage.Storer, 0)
+
+	defer func() {
+		// cleanup
+		if err != nil {
+			log.Error("create meta store", "error", err.Error())
+			for _, storer := range successfullyCreatedStorers {
+				_ = storer.DestroyUnit()
+			}
+			if shardHdrHashNonceUnits != nil {
+				for i := uint32(0); i < psf.shardCoordinator.NumberOfShards(); i++ {
+					if shardHdrHashNonceUnits[i] != nil {
```

Comment:
I would append these storers to the successfullyCreatedStorers as well, then there would be only one for loop over its range.

Reply:
done